### DISCUSSION TITLE
Unleash Proxy - Java Client

### DISCUSSION THREAD

**Discussion by nickcaplan** [Unleash/unleash#1798] [General]
Hello,

Is there a Java Client for the Unleash Proxy? I can't find one and wondering if this is deliberate and if so why?

We have access to Unleash through GitLab and want to ensure, by using the proxy, that we are protected from any GitLab outages, as I understand that the proxy will perform a certain amount of caching (if configured).

Any help greatly appreciated!

**Comment by thomasheartman** [comment]
Hey, @nickcaplan 👋🏼  

You're right! There's no proxy client for Java and that _is_ deliberate. I'll explain why in a moment, but just want to clarify one thing first:

All Unleash clients perform local run-time caching and a lot of them, including the Java client, also perform backups. So even if GitLab has an outage while your app is running, it'll keep working just fine with the last known state of all your features. The Java client also supports [bootstrapping](https://github.com/Unleash/unleash-client-java/#bootstrapping) and [local backups](https://github.com/Unleash/unleash-client-java/#local-backup). This should allow you to also start new clients even if GitLab's unleash instance can't be reached.

^ Would that solve your problems? ☺️ 

Now, regarding proxy clients: The proxy is primarily intended as a go-between between the Unleash server and client-side apps. Because Unleash evaluates all toggles on the client, the toggle data may contain sensitive data that you wouldn't want in the hands of your users or toggles that you don't want to expose. You also probably don't want to leak API keys to your users either. The proxy solves this by evaluating the toggles and exposing an endpoint that only returns a small amount of info on _enabled_ toggles.

In fact, the Proxy itself is just a thin wrapper around the Node.js client, so it performs the same amount of caching as that client does.

So in summary: you most likely don't need a proxy client for Java, as the client itself should handle all your needs if I understand correctly. Still, there may be some cases where a proxy client would be the right answer, but they're quite rare. How does that sound to you?

**Reply by chriswk** [reply]
Though, I do say you could probably use the [unleash-android-proxy-sdk](https://github.com/Unleash/unleash-android-proxy-sdk) as a java proxy client if you don't want to write one yourself. There's nothing depending on android in the code base, but that was it's intended use.
